### PR TITLE
network: re-design request queue

### PR DESCRIPTION
**Pull request by yuwata**
Split out of #22557.

### CODE REVIEW

File: src/network/networkd-queue.c [hunk 1]
```diff
@@ -313,9 +313,7 @@ int link_queue_request(
                       REQUEST_TYPE_DHCP4_CLIENT,
                       REQUEST_TYPE_DHCP6_CLIENT,
                       REQUEST_TYPE_NDISC,
-                      REQUEST_TYPE_RADV,
-                      REQUEST_TYPE_TC_QDISC,
-                      REQUEST_TYPE_TC_CLASS) ||
```
Comment (bluca) [inline src/network/networkd-queue.c:318]: this change is part of the `network: make address_configure() and friends take Request object` commit, but it doesn't seem related?

File: src/network/netdev/netdev.c [hunk 2]
```diff
@@ -632,16 +632,12 @@ static int netdev_is_ready_to_create(NetDev *netdev, Link *link) {
...
 
-int request_process_stacked_netdev(Request *req) {
-        NetDev *netdev;
-        Link *link;
+int stacked_netdev_process_request(Request *req, Link *link, void *userdata) {
+        NetDev *netdev = ASSERT_PTR(userdata);
         int r;
 
         assert(req);
-        assert(req->type == REQUEST_TYPE_NETDEV_STACKED);
```
Comment (bluca) [inline src/network/netdev/netdev.c:641]: all these type checks are being removed as part of the parameters reshuffle, are they no longer necessary?

File: src/network/networkd-queue.c [hunk 3]
```diff
@@ -92,7 +92,7 @@ static Request *request_free(Request *req) {
         return mfree(req);
 }
 
-DEFINE_TRIVIAL_CLEANUP_FUNC(Request*, request_free);
+DEFINE_TRIVIAL_REF_UNREF_FUNC(Request, request, request_free);
```
Comment (bluca) [inline src/network/networkd-queue.c:30]: it would be nice to add some explanation to the commit message, saying _why_ it's being changed to refcounted

File: src/network/networkd-queue.c [hunk 4]
```diff
@@ -463,3 +461,38 @@ int manager_process_requests(sd_event_source *s, void *userdata) {
...
+                assert(*req->counter > 0);
+                (*req->counter)--;
+                req->counter = NULL; /* To prevent double decrement on free. */
+        }
+
+        if (req->link && IN_SET(req->link->state, LINK_STATE_FAILED, LINK_STATE_LINGER))
+                return 0;
+
+        if (req->netlink_handler)
+                r = req->netlink_handler(nl, m, req, req->link, req->object);
```
Comment (bluca) [inline src/network/networkd-queue.c]: `return req->...` and `return 0` below? and remove r

File: src/network/networkd-queue.h [hunk 5]
```diff
@@ -68,9 +72,15 @@ typedef struct Request {
                 void *object;
         };
         void *userdata;
-        unsigned *message_counter;
-        link_netlink_message_handler_t netlink_handler;
-} Request;
+
+        /* incremented when requested, decremented when request is completed or failed. */
+        unsigned *counter;
+        /* called in netlink handler, the 'counter' is decremented before this is called.
```
Comment (bluca) [inline src/network/networkd-queue.h:77]: It would be good to have this explanation in the commit message - it's a large change, and the functional part is only explained here, halfway through the commit

File: src/network/networkd-queue.c [hunk 6]
```diff
@@ -63,7 +64,17 @@ static void request_free_object(RequestType type, void *object) {
         case REQUEST_TYPE_ROUTING_POLICY_RULE:
                 routing_policy_rule_free(object);
                 break;
-        case REQUEST_TYPE_SET_LINK:
+        case REQUEST_TYPE_SET_LINK_ADDRESS_GENERATION_MODE:
```
Comment (bluca) [inline src/network/networkd-queue.c]: It would be good to have a few lines in the commit message explaining _why_ it's a good idea to merge these two

**Review by bluca — APPROVED** [review]
It would be good to have more details in the commit messages about why the restructuring is done. Apart from this and a few small things, looks fine to me.

File: src/network/networkd-queue.c (hunk 1 above)
Comment (yuwata) [reply src/network/networkd-queue.c:318]: It is not irrelevant, but I will split the commit into two.

**Comment by yuwata** [comment]
@bluca Thank you for the review. Added detailed explanation in each commit. Upgrading the green label.

**Comment by bluca** [comment]
Thank you, this helps a lot to better understand the changes
